goxdr/goxdr.1.md
% goxdr(1)
% David Mazieres
%

# NAME

goxdr - Go XDR compiler

# SYNOPSIS

goxdr [-b] [-o _output.go_] [-p _package_] [_file1.x_ [_file2.x_ ...]]

# DESCRIPTION

goxdr compiles an RFC4506 XDR interface file to a set of go data
structures that can be either marshaled to or unmarshaled from
standard XDR binary format or traversed for other purposes such as
pretty-printing.  It does not rely on go's reflection facilities, and
so can be used to special-case handling of different XDR typedefs that
represent identical go types.

goxdr-compiled XDR types map to the most intuitive go equivalent:
strings map to strings, pointers map to pointers, fixed-size arrays
map to arrays, and variable-length arrays map to slices, without new
type declarations that might complicating assignment.  E.g., the XDR
`typedef string mystring<32>` is just a string, and so can be assigned
from a string.  This does mean you can assign a string longer than 32
bytes, but length limits are rigorously enforced during both
marshaling and unmarshaling.

## Type representations

To be consistent with go's symbol export policy, all types, enum
constants, and struct/union fields defined in an XDR file are
capitalized in the corresponding go representation.  Base XDR types
are mapped to their equivalent go types as follows:

    XDR type        Go type     notes
    --------------  ---------   ------------------------------
    bool            bool        const (TRUE=true; FALSE=false)
    int             int32
    unsigned int    uint32
    hyper           int64
    unsigned hyper  uint64
    float           float32
    double          float64
    quadruple       float128    but float128 is not defined
    string<n>       string
    opaque<n>       []byte
    opaque[n]       [n]byte
    T*              *T          for any XDR type T
    T<n>            []T         for any XDR type T
    T[n]            [n]T        for any XDR type T

Each XDR `typedef` is compiled to a go type alias (`type Alias =
Original`).

Each XDR `enum` declaration compiles to a defined type whose
representation is an `int32`.  The constants of the enum are defined
as go constants of the new defined type.

XDR defines bools as equivalent to an `enum` with name identifiers
`TRUE` and `FALSE`.  Hence, goxdr introduces these names as aliases
for go's `true` and `false`.  Be sure to use the capitalized versions
in case statements of XDR source files so as to maintain compatibility
with other languages and implementations.

An XDR `struct` is compiled to a defined type represented as a go
struct containing each field of the XDR struct.

An XDR `union` is compiled to a data structure with one public field
for the discriminant and one method for each non-void "arm
declaration" (i.e., declaration in a case statement) that returns a
pointer to a value of the appropriate type.  There is no need to
initialize the union when setting the discriminant; changing its value
just causes the appropriate method to return a non-nil pointer.
Invoking the wrong method for the current discriminant value calls
panic.

As an example, the following XDR:

~~~~{.c}
enum myenum {
    tag1 = 1,
    tag2 = 2,
    tag3 = 3
};

union myunion switch (myenum discriminant) {
    case tag1:
        int one;
    case tag2:
        string two<>;
    default:
        void;
};
~~~~

compiles to this go code:

~~~~{.go}
type Myenum int32
const (
    Tag1 = Myenum(1)
    Tag2 = Myenum(2)
    Tag3 = Myenum(3)
)
func XDR_Myenum(x XDR, name string, v *Myenum) {...}

type Myunion struct {
    Discriminant Myenum
    ....
}
func (u *Myunion) One() *int32 {...}
func (u *Myunion) Two() *string {...}
func XDR_Myunion(x XDR, name string, v *Myunion) {...}
~~~~

## The XDR interface

For every type `T` generated by goxdr (where `T` is the capitalized go
type), including typedefs, goxdr generates a function

~~~~{.go}
func XDR_T(x XDR, name string, v *T) {...}
~~~~
that can be used to marshal, unmarshal, or otherwise traverse the data
structure.  The `name` argument has no effect for RFC4506-compliant
binary marshaling, and can safely be supplied as the empty string
`""`.  However, when traversing an XDR type for other purposes such as
pretty-printing, `name` will be set to the nested name of the field
(with components separated by period).

The argument `x` implements the XDR interface and determines what
XDR_T actually does (i.e., marshal or unmarshal).  It has the
following interface:

~~~~{.go}
type XdrType interface {
	XdrValue() interface{}
	XdrPointer() interface{}
}

type XDR interface {
	Marshal(name string, val XdrType)
	Sprintf(string, ...interface{}) string
}
~~~~

`Sprintf` is expected to be a copy of `fmt.Sprintf`.  However, XDR
back-ends that do not make use of the `name` argument (notably
marshaling to RFC4506 binary format) can save some overhead by
returning an empty string.  Hence, the following are the two sensible
implementations of `Sprintf`:

~~~~{.go}
func (xp *MyXDR1) Sprintf(f string, args ...interface{}) string {
	return fmt.Sprintf(f, args...)
}

func (xp *MyXDR2) Sprintf(f string, args ...interface{}) string {
	return ""
}
~~~~

`Marshal` is the method that actually does whatever work will be
applied to the data structure.  The second argument, `val`, will be
the go value that must be marshaled/unmarshaled.  To simplify data
structure traversal, the value is not always just a pointer to the
value to be marshaled--in some cases the value is cast to or wrapped
in a defined type that allows handling of many different types to be
collapsed together.  Specifically, here is the type of `val` depending
on what is being marshaled:

* For bool and all 32-bit numeric types (including the size of
  variable-length arrays), `val` is passed as a pointer implementing
  `XdrNum32`, which allows the value to be extracted and set as a
  `uint32`.

* For all 64-bit numeric types, `val` is passed as a pointer
  implementing `XdrNum64`, which allows the value to be extracted and
  set as a `uint64`.

* For `struct` and `union` types, `val` is just a pointer to the type
  being marshaled.  However, these types implement the `XdrAggregate`
  interface, which allows the `Marshal` method of `XDR` to call the
  `XdrMarshal(x XDR, name string)` method on `val` to recurse through
  all fields of `val`.

* `enum` types are also passed as a simple pointer to the underlying
  field, but `enum` types implement `XdrNum32` instead of
  `XdrAggregate`.

* Fixed-length arrays (other than `opaque[]`) are passed to `Marshal`
  one element at a time, so `Marshal` is never called on the whole
  array (or on a pointer to the whole array).

* Variable-length arrays (other than `opaque<>`) are passed first as a
  pointer to a defined type implementing the `XdrVec` and
  `XdrAggregate` interfaces.  If `Marshal` calls the `XdrMarshal`
  function (as for a `struct` or `union`), it recurses, first calling
  `Marshal` on a value of `XdrSize`, then on each element of the
  vector.

* Similar to variable-length arrays, pointers use a defined type that
  implements the `XdrPtr` and `XdrAggregate` interfaces.  When
  recursing, `Marshal` is called first on another defined type that
  implements `XdrUint32` (capable of containing the value 0 or 1 to
  indicate nil or value-present), then, if the pointer is non-nil, on
  the underlying value.

* `string` is passed as an `XdrString`, which also encodes the size
  bound of the string and implements the `XdrVarBytes` and `XdrBytes`
  interfaces.

* `opaque<>` is passed as an `XdrVecOpaque` structure, which also
  implements the `XdrVarBytes` and `XdrBytes` interfaces.

* `opaque[]` is passed as an `XdrArrayOpaque` (user-defined slice type
  pointing to the entire array).  This type implements `XdrBytes` but
  not `XdrVarBytes`.

For most types, the original type or a pointer to it can be retrieved
via the `XdrPointer()` and `XdrValue()` methods, which return an
`interface{}`.  Two exceptions are `XdrArrayOpaque` (for which
`XdrValue()` returns a slice and `XdrPointer` returns `nil`), and the
fake `bool` on which `Marshal` is called for a pointer type (which
bool supports `XdrValue()`, but returns `nil` from `XdrPointer()`).

## XDR functions

As previously mentioned, each (capitalized) type `T` output by goxdr
also has function `XDR_T`.  For types that are instances of
`XdrAggregate` (that is `struct` and `union` types, as well as
pointers and variable-length arrays), this function is a simple
wrapper around the `Marshal` method:

~~~~{.go}
func XDR_T(x XDR, name string, v *T) {
    x.Marshal(name, v)
}
~~~~

For other types, however, this generated function casts `v` to a more
convenient alternate type implementing the interfaces described in the
previous subsection.  As an example, the following function in the
pre-defined boilerplate casts an ordinary `*int32` into the defined
type `*XdrInt32` which implements the `XdrNum32` interface:

~~~~{.go}
func XDR_int32(x XDR, name string, v *int32) {
        x.Marshal(name, (*XdrInt32)(v))
}
~~~~

Note that an XDR `Marshal` method can use a type switch to
special-case certain types.  However, this does not work for
`typedefs`, which goxdr emits as type aliases rather than defined
types---i.e. "`type Alias = Original`" rather than "`type Alias
Original`".  However, the `XDR_Alias` function for such a typedef
checks for a method called `x.Marshal_Alias(x XDR, name string, v
*Alias)` and calls it instead of `x.Marshal` if it exists, allowing
code to differentiate type aliases.

XDR functions panic with type `XdrError` (a user-defined string) if
the input is invalid or a value is out of range.

## Pre-defined XDR types

The predefined types `XdrOut`, `XdrIn`, and `XdrPrint` implement the
`XDR` interface and perform RFC4506 binary marshaling, RFC4506 binary
unmarshaling, and pretty-printing, respectively.

~~~~{.go}
type XdrOut struct {
	Out io.Writer
}
type XdrIn struct {
	In io.Reader
}
type XdrPrint struct {
	Out io.Writer
}
~~~~

# OPTIONS

goxdr supports the following options:

`-help`
:	Print a brief usage message.

`-b`
:	goxdr outputs boilerplate code to assist in marshaling and
unmarshaling values.  Only one copy of this boilerplate should be
included in a package.  If you use goxdr to compile all XDR input
files to a single go file (the recommended usage), then you will get
only one copy of the boilerplate.  However, if you compile different
XDR files into different go files, you will need to specify `-b` with
each XDR input file to avoid including the boilerplate, then run goxdr
with no input files (`goxdr -o goxdr_boilerplate.go`) to get one copy
of the boilerplate.

`-o` _output.go_
:	Write the output to file _output.go_ instead of standard output.

`-p` _package_
:	Specify the package name to use for the generated code.  The
default is for the generated code to declare `package main`.

# EXAMPLES

To serialize a data structure of type `MyType`:

~~~~{.go}
func serialize_Mytype(val *MyType) []byte {
	buf := &bytes.Buffer{}
	XDR_MyType(&XdrOut{ buf }, "", val)
	return buf.Bytes()
}
~~~~

To serialize/unserialize an arbitrary instance of `XdrAggregate` (any
struct or union, but not simpler types such as integers and strings):

~~~~{.go}
func serialize(val XdrAggregate) []byte {
	buf := &bytes.Buffer{}
	val.XdrMarshal(&XdrOut{ buf }, "")
	return buf.Bytes()
}

func deserialize(val XdrAggregate, in []byte) (e error) {
	defer func() {
		switch i := recover().(type) {
		case nil:

goxdr/goxdr.go
package main

import (
	"flag"
	"fmt"
	"io"
	"io/ioutil"
	"os"
	"strconv"
	"strings"
)

//go:generate goyacc -o parse.go parse.y
//go:generate sh -c "sed -e 's!^//UNCOMMENT:!!' header.go.in > header.go"

var progname string

type emitter struct {
	syms *rpc_syms
	output strings.Builder
	footer strings.Builder
	emitted map[string]struct{}
}

type Emittable interface {
	emit(e *emitter)
}

func capitalize(s string) string {
	if len(s) > 0 && s[0] >= 'a' && s[0] <= 'z' {
		return string(s[0] &^ 0x20) + s[1:]
	}
	return s
}

func uncapitalize(s string) string {
	if len(s) > 0 && s[0] >= 'A' && s[0] <= 'Z' {
		return string(s[0] | 0x20) + s[1:]
	}
	return s
}

func underscore(s string) string {
	if s == "" || s[0] == '_' {
		return s
	}
	return "_" + s
}

func parseXDR(out *rpc_syms, file string) {
	src, err := ioutil.ReadFile(file)
	if err != nil {
		fmt.Fprintf(os.Stderr, "%s: %s\n", progname, err)
		out.Failed = true
		return
	}
	l := NewLexer(out, file, string(src))
	yyParse(l)
}

func (e *emitter) done(typ string) bool {
	if _, ok := e.emitted[typ]; ok {
		return true
	}
	e.emitted[typ] = struct{}{}
	return false
}

func (e *emitter) append(out interface{}) {
	var s string
	switch t := out.(type) {
	case string:
		s = t
	case fmt.Stringer:
		s = t.String()
	default:
		panic("emitter append non-String")
	}
	e.output.WriteString(s)
}

func (e *emitter) printf(str string, args ...interface{}) {
	fmt.Fprintf(&e.output, str, args...)
}

func (e *emitter) xappend(out interface{}) {
	var s string
	switch t := out.(type) {
	case string:
		s = t
	case fmt.Stringer:
		s = t.String()
	default:
		panic("emitter append non-String")
	}
	e.footer.WriteString(s)
}

func (e *emitter) xprintf(str string, args ...interface{}) {
	fmt.Fprintf(&e.footer, str, args...)
}

func (e *emitter) get_typ(context idval, d *rpc_decl) idval {
	if d.typ.getgo() == "" {
		d.typ.setlocal(underscore(context.getgo()) + "_" + d.id.getgo())
		*d.inline_decl.getsym() = d.typ
		e.emit(d.inline_decl)
	}
	return d.typ
}

func (e *emitter) decltype(context idval, d *rpc_decl) string {
	typ := e.get_typ(context, d)
	switch d.qual {
	case SCALAR:
		return typ.getgo()
	case PTR:
		return fmt.Sprintf("*%s", typ)
	case ARRAY:
		return fmt.Sprintf("[%s]%s", d.bound, typ)
	case VEC:
		return fmt.Sprintf("[]%s", typ)
	default:
		panic("emitter::decltype invalid qual_t")
	}
}

// With line-ending comment showing bound
func (e *emitter) decltypeb(context idval, d *rpc_decl) string {
	ret := e.decltype(context, d)
	if (d.qual == VEC || d.typ.getgo() == "string") && d.bound.getgo() != "" {
		return fmt.Sprintf("%s // bound %s", ret, d.bound.getgo())
	}
	return ret
}

func (e *emitter) gen_ptr(typ idval) string {
	ptrtyp := "XdrPtr_" + typ.getgo()
	if typ.getgo()[0] == '_' {
		ptrtyp = "_XdrPtr" + typ.getgo()
	}
	if e.done(ptrtyp) {
		return ptrtyp
	}
	frag :=
`type $PTR struct {
	p **$TYPE
}
type _ptrflag_$TYPE $PTR
func (v _ptrflag_$TYPE) String() string {
	if *v.p == nil {
		return "nil"
	}
	return "non-nil"
}
func (v _ptrflag_$TYPE) GetU32() uint32 {
	if *v.p == nil {
		return 0
	}
	return 1
}
func (v _ptrflag_$TYPE) SetU32(nv uint32) {
	switch nv {
	case 0:
		*v.p = nil
	case 1:
		if *v.p == nil {
			*v.p = new($TYPE)
		}
	default:
		xdrPanic("*$TYPE present flag value %d should be 0 or 1", nv)
	}
}
func (v _ptrflag_$TYPE) XdrPointer() interface{} { return nil }
func (v _ptrflag_$TYPE) XdrValue() interface{} { return v.GetU32() != 0 }
func (v _ptrflag_$TYPE) XdrBound() uint32 { return 1 }
func (v $PTR) GetPresent() bool { return *v.p != nil }
func (v $PTR) SetPresent(present bool) {
	if !present {
		*v.p = nil
	} else if *v.p == nil {
		*v.p = new($TYPE)
	}
}
func (v $PTR) XdrMarshalValue(x XDR, name string) {
	if *v.p != nil {
		XDR_$TYPE(x, name, *v.p)
	}
}
func (v $PTR) XdrMarshal(x XDR, name string) {
	x.Marshal(name, _ptrflag_$TYPE(v))
	v.XdrMarshalValue(x, name)
}
func (v $PTR) XdrPointer() interface{} { return v.p }
func (v $PTR) XdrValue() interface{} { return *v.p }
`
	frag = strings.Replace(frag, "$PTR", ptrtyp, -1)
	frag = strings.Replace(frag, "$TYPE", typ.getgo(), -1)
	e.footer.WriteString(frag)
	return ptrtyp
}

// Return (bound, sbound) where bound is just the normalized bound,
// and sbound is a name suitable for embedding in strings (contains
// the more readable "unbounded" rather than 4294967295).
func (e *emitter) get_bound(b0 idval) (string, string) {
	b := b0
	for loop := map[string]bool{}; !loop[b.getx()]; {
		loop[b.getx()] = true
		if s, ok := e.syms.SymbolMap[b.getx()]; !ok {
			break
		} else if d2, ok := s.(*rpc_const); !ok {
			break
		} else {
			b = d2.val
		}
	}
	if b.getx() == "" {
		b.setlocal("0xffffffff")
	}
	if i32, err := strconv.ParseUint(b.getx(), 0, 32); err == nil {
		b.setlocal(fmt.Sprintf("%d", i32))
	} else if b.getx()[0] == '-' {
		// will make go compiler report the error
		return b0.getx(), b0.getx()
	}
	if b.getx() == "4294967295" {
		return b.getx(), "unbounded"
	}
	return b.getx(), b.getx()
}

func (e *emitter) gen_vec(typ, bound0 idval) string {
	bound, sbound := e.get_bound(bound0)
	vectyp := "XdrVec_" + sbound + "_" + typ.getgo()
	if typ.getgo()[0] == '_' {
		// '_' starts inline declarations, so only one size
		vectyp = "_XdrVec" + typ.getgo()
	}
	if e.done(vectyp) {
		return vectyp
	}
	frag :=
`type $VEC []$TYPE
func (v *$VEC) XdrBound() uint32 {
	const bound uint32 = $BOUND // Force error if not const or doesn't fit
	return bound
}
func (*$VEC) XdrCheckLen(length uint32) {
	if length > uint32($BOUND) {
		xdrPanic("$VEC length %d exceeds bound $BOUND", length)
	} else if int(length) < 0 {
		xdrPanic("$VEC length %d exceeds max int", length)
	}
}
func (v *$VEC) GetVecLen() uint32 { return uint32(len(*v)) }
func (v *$VEC) SetVecLen(length uint32) {
	v.XdrCheckLen(length)
	if int(length) <= cap(*v) {
		if int(length) != len(*v) {
			*v = (*v)[:int(length)]
		}
		return
	}
	newcap := 2*cap(*v)
	if newcap < int(length) { // also catches overflow where 2*cap < 0
		newcap = int(length)
	} else if bound := uint($BOUND); uint(newcap) > bound {
		if int(bound) < 0 {
			bound = ^uint(0) >> 1
		}
		newcap = int(bound)
	}
	nv := make([]$TYPE, int(length), newcap)
	copy(nv, *v)
	*v = nv
}
func (v *$VEC) XdrMarshalN(x XDR, name string, n uint32) {
	v.XdrCheckLen(n)
	for i := 0; i < int(n); i++ {
		if (i >= len(*v)) {
			v.SetVecLen(uint32(i+1))
		}
		XDR_$TYPE(x, x.Sprintf("%s[%d]", name, i), &(*v)[i])
	}
	if int(n) < len(*v) {
		*v = (*v)[:int(n)]
	}
}
func (v *$VEC) XdrMarshal(x XDR, name string) {
	size := XdrSize{ size: uint32(len(*v)), bound: $BOUND }
	x.Marshal(name, &size)
	v.XdrMarshalN(x, name, size.size)
}
func (v *$VEC) XdrPointer() interface{} { return (*[]$TYPE)(v) }
func (v *$VEC) XdrValue() interface{} { return ([]$TYPE)(*v) }
`
	frag = strings.Replace(frag, "$VEC", vectyp, -1)
	frag = strings.Replace(frag, "$TYPE", typ.getgo(), -1)
	frag = strings.Replace(frag, "$BOUND", bound, -1)
	e.footer.WriteString(frag)
	return vectyp
}

func (e *emitter) xdrgen(target, name string, context idval,
	d *rpc_decl) string {
	typ := e.get_typ(context, d)
	var frag string
	switch d.qual {
	case SCALAR:
		if typ.getgo() == "string" {
			frag = "\tx.Marshal($NAME, XdrString{$TARGET, $BOUND})\n"
		} else {
			frag = "\tXDR_$TYPE(x, $NAME, $TARGET)\n"
		}
	case PTR:
		ptrtype := e.gen_ptr(typ)
		frag = fmt.Sprintf("\tx.Marshal($NAME, %s{$TARGET})\n", ptrtype)
	case ARRAY:
		if typ.getgo() == "byte" {
			frag = "\tx.Marshal($NAME, XdrArrayOpaque((*$TARGET)[:]))\n"
			break;
		}
		frag =
`	for i := 0; i < len(*$TARGET); i++ {
			XDR_$TYPE(x, x.Sprintf("%s[%d]", $NAME, i), &(*$TARGET)[i])
		}
`
	case VEC:
		if typ.getgo() == "byte" {
			frag = "\tx.Marshal($NAME, XdrVecOpaque{$TARGET, $BOUND})\n"
			break;
		}
		vectyp := e.gen_vec(typ, d.bound)
		frag = fmt.Sprintf("\tx.Marshal($NAME, (*%s)($TARGET))\n", vectyp)
	}
	normbound := d.bound.getgo()
	if normbound == "" {
		normbound = "0xffffffff"
	}
	if len(target) >= 1 && target[0] == '&' {
		frag = strings.Replace(frag, "*$TARGET", target[1:], -1)
	}
	frag = strings.Replace(frag, "$TARGET", target, -1)
	frag = strings.Replace(frag, "$NAME", name, -1)
	frag = strings.Replace(frag, "$BOUND", normbound, -1)
	frag = strings.Replace(frag, "$TYPE", typ.getgo(), -1)
	return frag
}

func (e *emitter) emit(sym rpc_sym) {
	sym.(Emittable).emit(e)
}

func (r *rpc_const) emit(e *emitter) {
	e.printf("const %s = %s\n", r.id, r.val)
}

func (r0 *rpc_typedef) emit(e *emitter) {
	r := (*rpc_decl)(r0)
	e.printf("type %s = %s\n", r.id, e.decltypeb(gid(""), r))
	e.xprintf(
`func XDR_%[1]s(x XDR, name string, v *%[1]s) {
	if xs, ok := x.(interface{
		Marshal_%[1]s(string, *%[1]s)
	}); ok {
		xs.Marshal_%[1]s(name, v)
	} else {
	%[2]s	}
}
`, r.id, e.xdrgen("v", "name", gid(""), r))
}

func (r *rpc_enum) emit(e *emitter) {
	out := &strings.Builder{}
	fmt.Fprintf(out, "type %s int32\nconst (\n", r.id)
	for _, tag := range r.tags {
		fmt.Fprintf(out, "\t%s = %s(%s)\n", tag.id, r.id, tag.val)
	}
	fmt.Fprintf(out, ")\n")
	e.append(out)
	out.Reset()
	fmt.Fprintf(out,
`func XDR_%s(x XDR, name string, v *%[1]s) {
	x.Marshal(name, v)
}
`, r.id)
	fmt.Fprintf(out, "var _XdrNames_%s = map[int32]string{\n", r.id)
	for _, tag := range r.tags {
		fmt.Fprintf(out, "\tint32(%s): \"%s\",\n", tag.id, tag.id.getx())
	}
	fmt.Fprintf(out, "}\n")
	fmt.Fprintf(out, "var _XdrValues_%s = map[string]int32{\n", r.id)
	for _, tag := range r.tags {
		fmt.Fprintf(out, "\t\"%s\": int32(%s),\n", tag.id.getx(), tag.id)
	}
	fmt.Fprintf(out, "}\n")
	fmt.Fprintf(out,
`func (*%[1]s) XdrEnumNames() map[int32]string {
	return _XdrNames_%[1]s
}
func (v *%[1]s) String() string {
	if s, ok := _XdrNames_%[1]s[int32(*v)]; ok {
		return s
	}
	return fmt.Sprintf("%[1]s#%%d", *v)
}
func (v *%[1]s) Scan(ss fmt.ScanState, _ rune) error {
	if tok, err := ss.Token(true, xdrSymChar); err != nil {
		return err
	} else {
		stok := string(tok)
		if val, ok := _XdrValues_%[1]s[stok]; ok {
			*v = %[1]s(val)
			return nil
		}
		return XdrError(fmt.Sprintf("%%s is not a valid %[1]s.", stok))
	}
}
func (v *%[1]s) GetU32() uint32 {
	return uint32(*v)
}
func (v *%[1]s) SetU32(n uint32) {
	*v = %[1]s(n)
}
func (v *%[1]s) XdrPointer() interface{} {
	return v
}
func (v *%[1]s) XdrValue() interface{} {
	return *v
}
`, r.id)
	e.xappend(out)
}

func (r *rpc_struct) emit(e *emitter) {
	out := &strings.Builder{}
	fmt.Fprintf(out, "type %s struct {\n", r.id)
	for i := range r.decls {
		fmt.Fprintf(out, "\t%s %s\n", r.decls[i].id,
			e.decltypeb(r.id, &r.decls[i]))
	}
	fmt.Fprintf(out, "}\n")
	e.append(out)
	out.Reset()

	fmt.Fprintf(out,
`func (v *%[1]s) XdrPointer() interface{} {
	return v
}
func (v *%[1]s) XdrValue() interface{} {
	return *v
}
func (v *%[1]s) XdrMarshal(x XDR, name string) {
	if name != "" {
		name = x.Sprintf("%%s.", name)
	}
`, r.id)
	for i := range r.decls {
		out.WriteString(e.xdrgen("&v." + r.decls[i].id.getgo(),
			`x.Sprintf("%s` + r.decls[i].id.getx() + `", name)`,
			r.id, &r.decls[i]))
	}
	fmt.Fprintf(out, "}\n")
	fmt.Fprintf(out, "func XDR_%s(x XDR, name string, v *%s) {\n" +
		"\tx.Marshal(name, v)\n" +
		"}\n", r.id, r.id)
	e.xappend(out)
}

func (u *rpc_ufield) joinedCases() string {
	ret := u.cases[0].getgo()
	for i := 1; i < len(u.cases); i++ {
		ret += ", "
		ret += u.cases[i].getgo()
	}
	return ret
}

func (u *rpc_ufield) isVoid() bool {
	return u.decl.id.getx() == "" || u.decl.typ.getx() == "void"
}

func (r *rpc_union) emit(e *emitter) {
	out := &strings.Builder{}
	fmt.Fprintf(out, "type %s struct {\n", r.id)
	fmt.Fprintf(out, "\t%s %s\n", r.tagid, r.tagtype)
	for i := range r.fields {
		u := &r.fields[i]
		if u.isVoid() {
			continue
		}
		if (u.hasdefault) {
			fmt.Fprintf(out, "\t// default:\n")
		} else {
			fmt.Fprintf(out, "\t// %s:\n", u.joinedCases())
		}
		fmt.Fprintf(out, "\t//    %s() *%s\n", u.decl.id,
			e.decltypeb(r.id, &u.decl))
	}
	fmt.Fprintf(out, "\t_u interface{}\n" +
		"}\n")
	e.append(out)
	out.Reset()
	for i := range r.fields {
		u := &r.fields[i]
		if u.isVoid() {
			continue
		}
		ret := e.decltype(r.id, &u.decl)
		fmt.Fprintf(out, "func (u *%s) %s() *%s {\n", r.id, u.decl.id, ret)
		goodcase := fmt.Sprintf(
`		if v, ok := u._u.(*%[1]s); ok {
			return v
		} else {
			var zero %[1]s
			u._u = &zero
			return &zero
		}
`, ret)
		badcase := fmt.Sprintf(
`		xdrPanic("%s.%s accessed when %s == %%v", u.%[3]s)
		return nil
`, r.id, u.decl.id, r.tagid)
		fmt.Fprintf(out, "\tswitch u.%s {\n", r.tagid)
		if u.hasdefault && len(r.fields) > 1 {
			needcomma := false
			fmt.Fprintf(out, "\tcase ")
			for j := range r.fields {
				u1 := &r.fields[j]
				if r.hasdefault {
					continue
				}
				if needcomma {
					fmt.Fprintf(out, ",")
				} else {
					needcomma = true
				}
				fmt.Fprintf(out, "%s", u1.joinedCases())
			}
			fmt.Fprintf(out, ":\n%s\tdefault:\n%s", badcase, goodcase)
		} else {
			if u.hasdefault {
				fmt.Fprintf(out, "default:\n")
			} else {
				fmt.Fprintf(out, "\tcase %s:\n", u.joinedCases())
			}
			fmt.Fprintf(out, "%s", goodcase)
			if !u.hasdefault {
				fmt.Fprintf(out, "\tdefault:\n%s", badcase)
			}
		}
		fmt.Fprintf(out, "\t}\n")
		fmt.Fprintf(out, "}\n")
	}

	fmt.Fprintf(out, "func (u *%s) XdrValid() bool {\n", r.id)
	if r.hasdefault {
		fmt.Fprintf(out, "\treturn true\n")
	} else {
		fmt.Fprintf(out, "\tswitch u.%s {\n" + "\tcase ", r.tagid)
		needcomma := false
		for j := range r.fields {
			u1 := &r.fields[j]
			if needcomma {
				fmt.Fprintf(out, ",")
			} else {
				needcomma = true
			}
			fmt.Fprintf(out, "%s", u1.joinedCases())
		}
		fmt.Fprintf(out, ":\n\t\treturn true\n\t}\n\treturn false\n")
	}
	fmt.Fprintf(out, "}\n")

	fmt.Fprintf(out, "func (u *%s) XdrUnionTag() interface{} {\n" +
		"\treturn &u.%s\n}\n", r.id, r.tagid)
	fmt.Fprintf(out, "func (u *%s) XdrUnionTagName() string {\n" +
		"\treturn \"%s\"\n}\n", r.id, r.tagid)

	fmt.Fprintf(out, "func (u *%s) XdrUnionBody() interface{} {\n" +
		"\tswitch u.%s {\n", r.id, r.tagid)
	for i := range r.fields {
		u := &r.fields[i]
		if u.hasdefault {
			fmt.Fprintf(out, "\tdefault:\n")
		} else {
			fmt.Fprintf(out, "\tcase %s:\n", u.joinedCases())
		}
		if u.isVoid() {
			fmt.Fprintf(out, "\t\treturn nil\n")
		} else {
			fmt.Fprintf(out, "\t\treturn u.%s()\n", u.decl.id)
		}
	}
	fmt.Fprintf(out, "\t}\n")
	if !r.hasdefault {
		fmt.Fprintf(out, "\treturn nil\n")
	}
	fmt.Fprintf(out, "}\n")

	fmt.Fprintf(out, "func (u *%s) XdrUnionBodyName() string {\n" +
		"\tswitch u.%s {\n", r.id, r.tagid)
	for i := range r.fields {
		u := &r.fields[i]
		if u.hasdefault {
			fmt.Fprintf(out, "\tdefault:\n")
		} else {
			fmt.Fprintf(out, "\tcase %s:\n", u.joinedCases())
		}
		if u.isVoid() {
			fmt.Fprintf(out, "\t\treturn \"\"\n")
		} else {
			fmt.Fprintf(out, "\t\treturn \"%s\"\n", u.decl.id)
		}
	}
	fmt.Fprintf(out, "\t}\n")
	if !r.hasdefault {
		fmt.Fprintf(out, "\treturn \"\"\n")
	}
	fmt.Fprintf(out, "}\n")

	fmt.Fprintf(out,
`func (v *%[1]s) XdrPointer() interface{} {
	return v
}
func (v *%[1]s) XdrValue() interface{} {
	return *v
}
func (v *%[1]s) XdrMarshal(x XDR, name string) {
	if name != "" {
		name = x.Sprintf("%%s.", name)
	}
	XDR_%[2]s(x, x.Sprintf("%%s%[4]s", name), &v.%[3]s)
	switch v.%[3]s {
`, r.id, r.tagtype, r.tagid, r.tagid.getx())
	for i := range r.fields {
		u := &r.fields[i]
		if u.hasdefault {
			fmt.Fprintf(out, "\tdefault:\n")
		} else {
			fmt.Fprintf(out, "\tcase %s:\n", u.joinedCases())
		}
		if !u.isVoid() {
			out.WriteString(e.xdrgen("v." + u.decl.id.getgo() + "()",
				`x.Sprintf("%s` + u.decl.id.getx() + `", name)`,
				r.id, &u.decl))
		}
		out.WriteString("\t\treturn\n")
	}
	fmt.Fprintf(out, "\t}\n")
	if !r.hasdefault {
		fmt.Fprintf(out,
`	xdrPanic(fmt.Sprintf("invalid %[1]s (%%v) in %[2]s", v.%[1]s))
`, r.tagid, r.id)
	}
	fmt.Fprintf(out, "}\n")

	fmt.Fprintf(out, "func XDR_%s(x XDR, name string, v *%s) {\n" +
		"\tx.Marshal(name, v)\n" +
		"}\n", r.id, r.id)

	e.xappend(out)
}

func (r *rpc_program) emit(e *emitter) {
	// Do something?
}

func emitAll(syms *rpc_syms) string {
	e := emitter{
		syms: syms,
		emitted: map[string]struct{}{},
	}

	e.append(`
//
// Data types defined in XDR file
//
`)

	for _, s := range syms.Symbols  {
		e.append("\n")
		e.emit(s)
	}
	e.append(`
//
// Helper types and generated marshaling functions
//

`)
	e.append(e.footer.String())
	e.footer.Reset()
	return e.output.String()
}

func main() {
	opt_nobp := flag.Bool("b", false, "Suppress initial boilerplate code")
	opt_output := flag.String("o", "", "Name of output file")
	opt_pkg := flag.String("p", "main", "Name of package")
	opt_help := flag.Bool("help", false, "Print usage")
	progname = os.Args[0]
	if pos := strings.LastIndexByte(progname, '/'); pos >= 0 {
		progname = progname[pos+1:]
	}
	flag.Usage = func() {
		fmt.Fprintf(flag.CommandLine.Output(),
			"Usage: %s FLAGS [file1.x file2.x ...]\n", progname)
		flag.PrintDefaults()
	}
	flag.Parse()
	if (*opt_help) {
		flag.CommandLine.SetOutput(os.Stdout)
		flag.Usage()
		return
	}

	var syms rpc_syms
	for _, arg := range flag.Args() {
		parseXDR(&syms, arg)
	}
	if syms.Failed {
		os.Exit(1)
	}
	code := emitAll(&syms)

	out := os.Stdout
	if (*opt_output != "") {
		var err error
		out, err = os.Create(*opt_output)
		if err != nil {
			fmt.Fprintf(os.Stderr, "%s: %s\n", progname, err.Error())
			os.Exit(1)
		}
	}

	// https://github.com/golang/go/issues/13560#issuecomment-288457920
	fmt.Fprintf(out, "// Code generated by %s", progname);
	if len(os.Args) >= 1 {
		for _, arg := range os.Args[1:] {
			fmt.Fprintf(out, " %s", arg)
		}
	}
	fmt.Fprintln(out, "; DO NOT EDIT.\n");

	if *opt_pkg != "" {
		fmt.Fprintf(out, "package %s\n", *opt_pkg)
	}
	if !*opt_nobp {
		io.WriteString(out, header)
	}

	io.WriteString(out, code)
}
